src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.AbstractAddDocumentUndoUnit.cs

```diff
@@ -0,0 +1,46 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.VisualStudio.OLE.Interop;
+
+namespace Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem
+{
+    internal partial class VisualStudioWorkspaceImpl
+    {
+        private abstract class AbstractAddDocumentUndoUnit : AbstractAddRemoveUndoUnit
+        {
+            protected readonly DocumentInfo DocumentInfo;
+            protected readonly SourceText Text;
+
+            protected AbstractAddDocumentUndoUnit(
+                VisualStudioWorkspaceImpl workspace,
+                DocumentInfo docInfo,
+                SourceText text)
+                : base(workspace, docInfo.Id.ProjectId)
+            {
+                DocumentInfo = docInfo;
+                Text = text;
+            }
+
+            public override void Do(IOleUndoManager pUndoManager)
+            {
+                var currentSolution = Workspace.CurrentSolution;
+                var fromProject = currentSolution.GetProject(FromProjectId);
+
+                if (fromProject != null)
+                {
+                    var updatedProject = AddDocument(fromProject);
+                    Workspace.TryApplyChanges(updatedProject.Solution);
```

Comment:
So are we putting new undo units onto the stack here while we're processing an undo? Is that safe and/or expected in this context? Put another way, if you undo this, what does the redo stack show? Does that work as expected?

I can't help but to notice we're violating the rule though as documented in IOleUndoUnit::Do, at least strictly:

> If pUndoManager is not NULL, then the unit is required to put a corresponding unit on the redo or undo stack. As a result, this method either moves itself to the redo or undo stack, or it creates a new undo unit and adds it to the appropriate stack. After creating a new undo unit, this undo unit calls IOleUndoManager::Open or IOleUndoManager::Add. The undo manager will put the new undo unit on the undo or redo stack depending on its current state.

The text editors abstraction for undo/redo has units have both an "undo" and "do it again" action and it lets the units move back and forth. I'm not sure if we need to do that or not...